Update to use Thrust in rocSPARSE examples and tests

BenBrock opened the pull request:
This PR applies some minor updates to @yhmtsai's PR:

1. Update [tests](https://github.com/SparseBLAS/spblas-reference/blob/dev/brock/rocsparse_spmv/test/gtest/rocsparse/spmv_test.cpp) and [examples](https://github.com/SparseBLAS/spblas-reference/blob/dev/brock/rocsparse_spmv/examples/rocsparse_examples/simple_spmv.cpp) to use Thrust.
2. Re-write [HIP memory allocator](https://github.com/SparseBLAS/spblas-reference/blob/dev/brock/rocsparse_spmv/include/spblas/vendor/rocsparse/hip_allocator.hpp) to follow C++ conventions, support asynchronous allocation.
3. Add compilation instructions to the README.

I think once we apply this update and get the CI working (currently in progress), @yhmtsai's PR #41 will be ready to merge.

spencerpatty commented on README.md:
```diff
@@ -104,7 +104,8 @@ brock@slothius:~/src/spblas-reference$ CXX=g++-13 cmake -B build
 ### Compiling with a Vendor Backend
 A vendor backend can be enabled by passing in an `-DENABLE_{BACKEND}=ON` switch
-to `cmake`.  Currently, oneMKL is the only supported backend.
+to `cmake`.  Currently, oneMKL, ArmPL, and rocSPARSE are the supported vendor
+backends.
```
nice update, we haven't been diligent about looking at docs with each extension of backends, but maybe should ...

I have thought about trying to write up a "contribution.md" or PR template checklist for helping build our development culture

BenBrock replied:
Yeah, I think adding a `contributions.md` would be nice.  Currently there's just a section in the `README.md`.

spencerpatty commented on README.md:
```diff
@@ -118,6 +119,28 @@ brock@slothius:~/src/spblas-reference$ source /opt/intel/oneapi/setvars.sh
 brock@slothius:~/src/spblas-reference$ CXX=icpx cmake -B build -DENABLE_ONEMKL_SYCL=ON
```
you might want to check this doc for any IP leaks like machine names or userids and whatnot that maybe aren't necessary here :)

BenBrock replied:
Yeah, I've checked over things—no ID leaks.  (Slothius is a fake system name...)

spencerpatty commented on examples/CMakeLists.txt:
```diff
@@ -3,9 +3,14 @@ function(add_example example_name)
   target_link_libraries(${example_name} spblas fmt)
 endfunction()
 
+if (NOT ENABLE_ROCSPARSE)
```
thinking about future with at least Nvidia, AMD and Intel backends that will be non-cpu and might need the special subdirectories of examples, how will we want this to scale ?  will the "rocsparse_examples" name be extended, or will there be another folder for each "device" ?  or should we have a common "device_examples" folder which uses Thrust for AMD/NVIDIA  and some other similar construct for MKL SYCL ?  or do we keep them separate ?

BenBrock replied:
In the future, I think we should have `examples/device`, which all GPU backends build using Thrust (including oneMKL SYCL if we provide a minimal Thrust interface like we talked about, @spencerpatty).  The GPU backends can have their own additional examples if they like (e.g. `examples/rocsparse`, `examples/onemkl`) that are written however they like (perhaps using HIP instead of Thrust).

spencerpatty replied:
might be possible to use gcc for the app compiler with sycl backend, but would need icx -fsycl for compiling any sycl specific code at least to spir-v ...

BenBrock replied:
Yeah, oneMKL SYCL is not set up to compile with a compiler other than `icpx`, right?  From what I understand, we'd need to explicitly instantiate all the `multiply` algorithms for each set of template parameters we want, separately compile them using `icpx`, and then link to a program that uses another compiler.

spencerpatty replied:
so if I remember correctly, it is possibly to use ` icpx -fsycl ... -fsycl-link ...` to compile the sycl specific device objects (and it does any work for sycl -> spirv module compilation that is usually done at link stage earlier at the device object creation stage) and then a separate linker like gcc or g++ could be used to link the entire application of other objects and the device objects ...  

https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2023-0/use-third-party-compiler-as-host-compiler-dpc-code.html

spencerpatty commented on examples/rocsparse_examples/simple_spmv.cpp:
```diff
@@ -0,0 +1,66 @@
+  auto&& [values, rowptr, colind, shape, nnz] =
+      generate_csr<T, spblas::index_t>(m, n, nnz_in);
+
+  thrust::device_vector<T> d_values(values);
```
I see use of `thrust::device_vector` and `thrust::copy` below.  Any other thrust capabilities we think might eventually be needed ?  I can think maybe of `memset` ?

BenBrock replied:
Yeah, we might possibly need `thrust::fill` (C++ equivalent of `memset`), although you can get the same behavior calling the `resize()` method on the vector.  (`v.resize(0)` is specified not to change capacity [it just resets the `size`] and `v.resize(n, value)` will fill in `values`.)

BenBrock replied:
I don't think there's anything else we need; we basically just need to allocate/deallocate memory and fill in values.

spencerpatty commented on include/spblas/vendor/rocsparse/exception.hpp:
```diff
@@ -5,43 +5,79 @@
 }
 
-// throw an exception if the rocsparse_status is not rocsparse_status_success.
+// Throw an exception if the rocsparse_status is not rocsparse_status_success.
 void throw_if_error(rocsparse_status error_code) {
```
should we have these capture __FILE__ and __LINE__ string details from level above, so it can also spit out which file the exception was thrown from ?

spencerpatty replied:
I suppose if we have a log_warning("") at each call, then that could be turned on and it would give stuff, but again it is the same issue:  throw_if_error() would need the right function/line number provided to it ... since the checks are inside of throw_if_error ...

BenBrock replied:
I'd be happy to add that in—we could have a `SPBLAS_ROCSPARSE_CHECK` macro that grabs the line number and generates the exception.  That would be useful for both rocSPARSE and cuSPARSE.  (oneMKL I assume we will just allow oneMKL's exceptions to carry through?  At some point we need to define what exceptions the Sparse BLAS algorithms can throw and when.)

BenBrock commented:
The rocSPARSE CI is now set up properly using ICL's resources ([24e0378](https://github.com/SparseBLAS/spblas-reference/commit/24e037807089998d356f54d6c9d875a2832e4687)).

yhmtsai commented on include/spblas/vendor/rocsparse/multiply.hpp:
```diff
@@ -107,7 +106,7 @@ class spmv_state_t {
-  std::shared_ptr<allocator> alloc_;
+  rocsparse::hip_allocator<char> alloc_;
   long unsigned int buffer_size_;
   void* workspace_;
```
```suggestion
  char* workspace_;
```
to avoid the casting

BenBrock replied:
Fixed.

yhmtsai commented on include/spblas/vendor/rocsparse/hip_allocator.hpp:
```diff
@@ -0,0 +1,74 @@
+    hipError_t error = hipMallocAsync(&ptr, size * sizeof(T), stream());
+    throw_if_failure(error);
+
+    return reinterpret_cast<T*>(ptr);
```
```suggestion
    return static_cast<T*>(ptr);
```

yhmtsai commented on include/spblas/vendor/rocsparse/multiply.hpp:
```diff
@@ -7,38 +7,37 @@
-  spmv_state_t() : spmv_state_t(std::make_shared<detail::rocm_allocator>()) {}
+  spmv_state_t() : spmv_state_t({}) {}
 
-  spmv_state_t(std::shared_ptr<allocator> alloc)
+  spmv_state_t(rocsparse::hip_allocator<char> alloc)
```
How does it work if user would like to have a memory pool shared in many state?

yhmtsai commented on include/spblas/vendor/rocsparse/hip_allocator.hpp:
```diff
@@ -0,0 +1,74 @@
+  bool operator!=(const hip_allocator&) const = default;
+
+  template <typename U>
+  struct rebind {
```
I think it is deleted in c++20

BenBrock replied:
It's removed from `std::allocator<T>`, but it's still in `std::allocator_traits<...>`, and you can still define it on your own allocators.

yhmtsai commented on include/spblas/vendor/rocsparse/hip_allocator.hpp:
```diff
@@ -0,0 +1,74 @@
+    using other = hip_allocator<U, Alignment>;
+  };
+
+  hipStream_t stream() const noexcept {
```
should we provide it or just a workaround? I thought we said it is belong to the execution policy

yhmtsai reviewed (commented):
gcc support claims?: it likely works but we can only test hipcc (kind of clang) now. How do we say about the gcc support for hip backend? It stops me adding rocThrust for the test.

I would like to keep the same test and example for every backend.
I can do it on my pr after merging this. 

I do not think the allocator we need to follow the std allocator. I feel that it raises the difficulty of user's allocator. Also, it follows the allocator design, but user can not use it to any std container.

spencerpatty commented on include/spblas/vendor/rocsparse/multiply.hpp:
```diff
@@ -7,38 +7,37 @@
   }
 
-  spmv_state_t(std::shared_ptr<allocator> alloc, rocsparse_handle handle)
+  spmv_state_t(rocsparse::hip_allocator<char> alloc, rocsparse_handle handle)
```
some have `rocsparse::`  namespace and others have `__rocsparse::` namespace.  IS this on purpose ?  Can we align on a single name ?  or is there a reason for having them be separate ? in which case, can we add something to the name to highlight that purpose?

BenBrock replied:
The convention I follow in the oneMKL and ArmPL backends is to place private, implementation-defined functions and classes in a `spblas::__backend_name_here` namespace (`spblas::__mkl` and `spblas::__armpl`, respectively).  Here I place implementation-only code inside the `spblas::__rocsparse` namespace.

Now, there's a question about whether `hip_allocator` should be public or not.  I'm not sure whether we want it to be; however, at some point we will probably want to provide _some_ allocators.  e.g., if we want to provide an allocator backed by a buffered memory pool.  Figuring that it might at some point be public, I put it in the namespace `spblas::rocsparse`, which is visible to the user.

In terms of a spec., I don't think we say anything about memory allocators being defined—those are all up to the backend vendor implementation.

spencerpatty replied:
I'm not sure I personally like having `__rocsparse::` and `rocsparse::` namespaces in the same file, but I get the idea of internal implementation vs public...  I would more expect to see something like `__rocsparse::impl::` vs `__rocsparse::`   but I can see that there is a pattern you are using and that is fine with me for now :)

BenBrock commented on examples/CMakeLists.txt:
```diff
@@ -3,9 +3,14 @@ function(add_example example_name)
 add_example(matrix_opt_example)
 add_example(spmm_csc)
+else()
+  set_source_files_properties(${example_name}.cpp PROPERTIES LANGUAGE HIP)
```
Just updated to create `examples/device`, which uses Thrust, and `examples/rocsparse`, which uses HIP only.  I would suggest this is the way to go forward, as it gives users both a unified device example that works on any of the GPU backends `examples/device` and rocSPARSE-specific examples.

This also solves the problem of testing both `hipcc` and `gcc`, since `examples/device` will be compiled with Thrust (and thus `hipcc`) while `examples/rocsparse` will be compiled with GCC (or any other host compiler you like).

I would propose that examples are enough to ensure that everything compiles with a host compiler—we basically want to ensure we don't introduce any `hipcc`-required components by accident, and compiling an example will ensure that.

Also fixed the typo you mentioned.

BenBrock commented on test/gtest/CMakeLists.txt:
```diff
@@ -13,6 +11,10 @@ else()
     triangular_solve_test.cpp
   )
+elseif(ENABLE_ROCSPARSE)
+  set_source_files_properties(rocsparse/spmv_test.cpp PROPERTIES LANGUAGE HIP)
```
See my comment above.

BenBrock commented on test/gtest/CMakeLists.txt:
```diff
@@ -13,6 +11,10 @@ else()
+elseif(ENABLE_ROCSPARSE)
+  set_source_files_properties(rocsparse/spmv_test.cpp PROPERTIES LANGUAGE HIP)
+  add_executable(spblas-tests
+                 rocsparse/spmv_test.cpp)
```
See my comment above.  We can have an `examples/device` set of examples that are unified.

BenBrock commented:
@yhmtsai I just updated the PR to answer most of your questions, I believe.

1. Unifying examples/tests across all backends: I've just modified the PR to store GPU examples in `examples/device`.  These examples will be compiled with `rocSPARSE`, `cuSPARSE`, and `oneMKL SYCL` backends.  (I've discussed with @spencerpatty writing a short Thrust stub in SYCL to provide the features we need.  I have most of the pieces already sitting around.)  We can do the same thing with tests.  (Although I think we can just wait until I get my cuSPARSE PR in.)
2. Compilation with a non-`hipcc` host compiler: I've added a separate directory for `rocsparse`-specific examples in `examples/rocsparse` (perhaps `examples/hip` would be better?).  These examples are compiled with the host compiler (GCC on our CI system), and everything is currently working.
3. There is the question of what interface the allocator should offer.  You are right that it is a bit wonky to have an allocator that follows the standard C++ allocator named requirement but whose pointers are not accessible on the host.  However, I believe there is some precedent for this in previous versions of Thrust.  Also, I'm not sure what interface to use if we don't use the C++ allocator interface—are we going to make up our own?  This allocator interface can also be internal-only for now; however, we'll eventually want the user to be able to provide their own allocator, so we need to come up with something.  I think using the C++ allocator interface is probably the most standard/reasonable thing to do.

Let me know your thoughts; I'd like to merge this into your PR and then into main in the next day or so so that I can begin working on a cuSPARSE PR.

yhmtsai reviewed (commented):
do not need to worry about the cuSPARSE. when we have some certain decision on rocSPARSE. I will update the cuSPARSE pr to fit the same guideline although I will still need your help on the ci setup for cuSPARSE.

1. I always consider CPU is another kind of accelerator. I would like to unify all not just gpu test in my PR. Paying another copy in the test if we unify the gpu test and cpu test is not a big downside to me. If you need to implement thrust-like for oneMKL, it should be something like our own array but follows more thrust interface, I guess? I do not see why we can not just use our own array especially for testing.

2. Thanks for updating related to non-hipcc concern. It is not the test, but it can make me more confident on different compiler support.  Will we need to add each function without Thrust for that? If yes, adding that directly to the test not example might be bette rbecause we eventually need to implement it. Thrust just introduces more work not reduce work.

3. I take your comment as we need to discuss more in the meeting about what the allocator looks like. IMO, user should only need to implement what we will use in the function. also, the pool allocator is the concern. I will say that we start from the simple rather than the complex one.

BenBrock commented:
1. CPU is an accelerator: in my opinion unifying CPU and GPU examples and tests introduces unnecessary complications into the CPU versions.  Not only will this probably require another copy on the CPU, but more importantly it would introduce unnecessary additional code that could confuse users.  I think having one set of examples/tests for CPUs and one set of examples/tests for GPUs is fine.

2. non-`hipcc` builds: I think the the idea here is to ensure that things can compile, but not to comprehensively test correctness.  We're already doing functional testing with Thrust, so I don't think there's a reason to re-write all the tests with HIP just to ensure things will compile.  I think having some/all examples in HIP will be enough to provide good compilation coverage, and more importantly provide a guide of proper usage for users.  We probably want that anyway, since not everyone will use Thrust.

3. I agree that the user should not _need_ to implement an allocator.  In the vast majority of cases, the default allocator provided by the vendor should work and provide good performance.  However, the user _should_ be able to provide an allocator if they want to.  (This is a requirement for certain environments, e.g. MATLAB.)  C++ allocators are the standard way to provide an allocator in C++, so I think it only makes sense that we use them rather than going off and inventing something on our own.

yhmtsai reviewed (approved):
1. It also means that we need to carefully ensure the minimal interface between CPU and GPU are the same.
2. if we do not need to implement all each function, then I am somehow okay with the current one. I have experienced some fundamental but critical issues like array indexing or division from different compiler such that I have more concerns regarding to the compiler.
3. not sure about it. Implementing it as standard way but giving the issue with the standard seems not great. Also, should it be ised as value or pointer?
I can go with it currently.
I think there are two good points to revisit this design: 1) pool allocator synced and shared between different operations. 2) exeuction policy

BenBrock commented:
Sounds good; I will merge this, and we can chat about the allocator/execution policy issues on Thursday.
